kotlinx-coroutines-core/common/src/channels/Produce.kt

```diff
@@ -25,6 +25,32 @@ public interface ProducerScope<in E> : CoroutineScope, SendChannel<E> {
     val channel: SendChannel<E>
 }
 
+/**
+ * Suspends the current coroutine until the channel is either [closed][SendChannel.close] or [cancelled][ReceiveChannel.cancel]
+ * and invokes the given [block] before resuming the coroutine.
+ *
+ * Example of usage:
+ * ```
+ * val callbackEventsStream = produce {
+ *     val disposable = registerChannelInCallback(channel)
+ *     await { disposable.dispose() }
+ * }
+ * ```
+ */
+@ExperimentalCoroutinesApi
+public suspend fun <T> ProducerScope<T>.await(block: () -> Unit = {}) {
```

Comment:
Design: It should be called `awaitClose` (consistently with `invokeOnClose`) or otherwise it is not clear what is being awaited when reading `await { disposable.dispose() }` code.

Reply:
Also, note that it is an extension on `ProducerScope` rather than `SendChannel`.